packages/aws-cdk/lib/api/hotswap-deployments.ts

```diff
@@ -54,27 +55,75 @@ export async function tryHotswapDeployment(
   return { noOp: hotswappableChanges.length === 0, stackArn: cloudFormationStack.stackId, outputs: cloudFormationStack.outputs, stackArtifact };
 }
 
-async function findAllHotswappableChanges(
-  stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
-): Promise<HotswapOperation[] | undefined> {
-  const promises = new Array<Promise<ChangeHotswapResult>>();
-  stackChanges.resources.forEachDifference(async (logicalId: string, change: cfn_diff.ResourceDifference) => {
-    promises.push(isHotswappableLambdaFunctionChange(logicalId, change, evaluateCfnTemplate));
-  });
-  return Promise.all(promises).then(hotswapDetectionResults => {
-    const hotswappableResources = new Array<HotswapOperation>();
-    let foundNonHotswappableChange = false;
-    for (const lambdaFunctionShortCircuitChange of hotswapDetectionResults) {
-      if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
-        foundNonHotswappableChange = true;
-      } else if (lambdaFunctionShortCircuitChange === ChangeHotswapImpact.IRRELEVANT) {
-        // empty 'if' just for flow-aware typing to kick in...
-      } else {
-        hotswappableResources.push(lambdaFunctionShortCircuitChange);
-      }
+async function findAllHotswappableChanges(stackChanges: cfn_diff.TemplateDiff, evaluateCfnTemplate: EvaluateCloudFormationTemplate):
+Promise<HotswapOperation[] | undefined> {
+  const hotswappableResources = new Array<HotswapOperation>();
+  let foundNonHotswappableChange = false;
+  const promises: Array<Array<Promise<ChangeHotswapResult>>> = [];
+
+  // gather the results of the detector functions
+  stackChanges.resources.forEachDifference((logicalId: string, change: cfn_diff.ResourceDifference) => {
+    const nonHotswappableResourceFound = isResourceChangeHotswappable(change);
+
+    if (nonHotswappableResourceFound === ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT) {
+      foundNonHotswappableChange = true;
+    } else if (nonHotswappableResourceFound === ChangeHotswapImpact.IRRELEVANT) {
+      // empty 'if' just for flow-aware typing to kick in...
+    } else {
+      const detectorResults = new Array<Promise<ChangeHotswapResult>>();
+      detectorResults.push(isHotswappableLambdaFunctionChange(logicalId, nonHotswappableResourceFound, evaluateCfnTemplate));
+      detectorResults.push(isHotswappableStateMachineChange(logicalId, nonHotswappableResourceFound, evaluateCfnTemplate));
+
+      promises.push(detectorResults);
     }
-    return foundNonHotswappableChange ? undefined : hotswappableResources;
   });
+
+  // resolve all detector results
+  for (const detectorResultPromises of promises) {
+    await Promise.all(detectorResultPromises).then(hotswapDetectionResults => {
```

Comment:
I really don't love this structure.

Can you create a new variable, something like `changesDetectionResults`, of type `Array<Array< ChangeHotswapResult>>`, and populate it by `await`ing on the `Promise.all()` result of iterating through `promises`, and then just run this logic on `changesDetectionResults`, without worrying about any asynchronous operations? I think it will make this code much clearer.